# 🚗Exploring Ebay Car Sales Data🚗
**(Used cars database : Over 370,000 used cars scraped from Ebay Kleinanzeigen)**

### 👉<u>Goal</u> -
* **Data Cleaning of Ebay Car Sales Data**
* **Data Analysis of the listed used car**

**👉[Dataset link](https://www.kaggle.com/orgesleka/used-cars-database/data) : Data is of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.**

🔧**<u>Tools used</u>: pandas, numpy, python**

**Over 370000 used cars scraped with Scrapy from Ebay-Kleinanzeigen.**
**📒Data Dictionary provided with data is as follows:**

* **dateCrawled** - When this ad was first crawled. All field-values are taken from this date.
* **name** - Name of the car.
* **seller** - Whether the seller is private or a dealer.
* **offerType** - The type of listing
* **price** - The price on the ad to sell the car.
* **abtest** - Whether the listing is included in an A/B test.
* **vehicleType** - The vehicle Type.
* **yearOfRegistration** - The year in which the car was first registered.
* **gearbox** - The transmission type.
* **powerPS** - The power of the car in PS.
* **model** - The car model name.
* **kilometer** - How many kilometers the car has driven.
* **monthOfRegistration** - The month in which the car was first registered.
* **fuelType** - What type of fuel the car uses.
* **brand** - The brand of the car.
* **notRepairedDamage** - If the car has a damage which is not yet repaired.
* **dateCreated** - The date on which the eBay listing was created.
* **nrOfPictures** - The number of pictures in the ad.
* **postalCode** - The postal code for the location of the vehicle.
* **lastSeenOnline** - When the crawler saw this ad last online.

The fields lastSeen and dateCreated could be used to estimate how long a car will be at least online before it is sold.

## ⭐Step1 - Import Libraries

In [1]:
import pandas as pd
import numpy as np

### ⭐Step2 - Read the Data

In [2]:
# Path of file
autos_file_path = (r'C:\Users\anuja\Downloads\autos.csv')
autos = pd.read_csv(autos_file_path, encoding='Latin-1')

✏️**Exploring dataset using feature of jupyter notebook is its ability to render the first few and last few values of any pandas object.**

In [3]:
autos

dateCrawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        seller offerType  price   abtest vehicleType  yearOfRegistration  \
0       privat   Angebot    480     test         NaN                1993   
1       privat   Angebot  18300     test       coupe                2011   
2       privat   Angebot   9800     test         suv                2004   
3       privat   Angebot   1500     test  kleinwagen                2001   
4       privat   Angebot   3600     test  kleinwagen                2008   
...        ...       ...    ...      ...         ...                 ...   
371523  privat   Angebot   2200     test         NaN                2005   
371524  privat   Angebot   1199     test      cabrio                2000   
371525  privat   Angebot   9200     test         bus                1996   
371526  privat   Angebot   3400     test       kombi                2002   
371527  privat   Angebot  28990  control   limousine                2013   

          gearbox  powerPS        model  kilometer  monthOfRegistration  \
0         manuell        0         golf     150000                    0   
1         manuell      190          NaN     125000                    5   
2       automatik      163        grand     125000                    8   
3         manuell       75         golf     150000                    6   
4         manuell       69        fabia      90000                    7   
...           ...      ...          ...        ...                  ...   
371523        NaN        0          NaN      20000                    1   
371524  automatik      101       fortwo     125000                    3   
371525    manuell      102  transporter     150000                    3   
371526    manuell      100         golf     150000                    6   
371527    manuell      320      m_reihe      50000                    8   

       fuelType           brand notRepairedDamage          dateCreated  \
0        benzin      volkswagen               NaN  2016-03-24 00:00:00   
1        diesel            audi                ja  2016-03-24 00:00:00   
2        diesel            jeep               NaN  2016-03-14 00:00:00   
3        benzin      volkswagen              nein  2016-03-17 00:00:00   
4        diesel           skoda              nein  2016-03-31 00:00:00   
...         ...             ...               ...                  ...   
371523      NaN  sonstige_autos               NaN  2016-03-14 00:00:00   
371524   benzin           smart              nein  2016-03-05 00:00:00   
371525   diesel      volkswagen              nein  2016-03-19 00:00:00   
371526   diesel      volkswagen               NaN  2016-03-20 00:00:00   
371527   benzin             bmw              nein  2016-03-07 00:00:00   

        nrOfPictures  postalCode             lastSeen  
0                  0       70435  2016-04-07 03:16:57  
1                  0       66954  2016-04-07 01:46:50  
2                  0       90480  2016-04-05 12:47:46  
3                  0       91074  2016-03-17 17:40:17  
4                  0       60437  2016-04-06 10:17:21  
...              ...         ...                  ..

✏️**Exploring dataset using info and head commands**

In [4]:
autos.info()
autos.head(2)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

dateCrawled                  name  seller offerType  price abtest  \
0  2016-03-24 11:52:17            Golf_3_1.6  privat   Angebot    480   test   
1  2016-03-24 10:58:45  A5_Sportback_2.7_Tdi  privat   Angebot  18300   test   

  vehicleType  yearOfRegistration  gearbox  powerPS model  kilometer  \
0         NaN                1993  manuell        0  golf     150000   
1       coupe                2011  manuell      190   NaN     125000   

   monthOfRegistration fuelType       brand notRepairedDamage  \
0                    0   benzin  volkswagen               NaN   
1                    5   diesel        audi                ja   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50

**✏️Checking Percentage of the Missing values in dataset**

In [5]:
# how many total missing values are there?
total_cells = np.product(autos.shape)
total_missing = autos.isnull().sum().sum()

# percent of data that is missing
(total_missing/total_cells) * 100

2.4763678646023988

**2% of data is missing in the dataset.**

### ⛳Observations 1- Read the Data Observations
* **Object type :** Object type representation of columns, which means they are represented by strings, not numbers.


* **Null Values :** 2% of data is missing in the dataset.eg- vehicleType, gearbox, etc.


* **Upper and Lowercase Letters :** eg - dateCrawled, offerType, etc.


* **No Underscores :** in column labels which makes them harder to read as  camelcase (instead of Python's preferred snakecase) is used.

  So in order to make them readable, we need to convert camelcase to snakecase and update the column labels


* **Updation of Column Labels:** eg- column price has recorded values in dollars and column names can be updated as price(in dollars)
 
 **Various Data Cleaning is to be done for effective analysis.**

### ⭐Step3 - Updating Column Labels

In [6]:
# To get columns
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [7]:
# To modify column labels
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price_$', 'ab_test',
       'vehicle_type', 'registration_year', 'gear_box', 'power_PS', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'num_photos', 'postal_code',
       'last_seen']
# to get dataset
autos.head(2)

date_crawled                  name  seller offer_type  price_$  \
0  2016-03-24 11:52:17            Golf_3_1.6  privat    Angebot      480   
1  2016-03-24 10:58:45  A5_Sportback_2.7_Tdi  privat    Angebot    18300   

  ab_test vehicle_type  registration_year gear_box  power_PS model  kilometer  \
0    test          NaN               1993  manuell         0  golf     150000   
1    test        coupe               2011  manuell       190   NaN     125000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   
1                   5    diesel        audi                ja   

            ad_created  num_photos  postal_code            last_seen  
0  2016-03-24 00:00:00           0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00           0        66954  2016-04-07 01:46:50

### ⛳Observations 2 - Updating Column Labels Observations
By now we have updated the column labels in the dataset by adding underscores to increase easy readibility and shortening of column labels for ease in handling.

### ⭐Step4 - Data Exploration

In [8]:
# to look at descriptive statistics for all columns.
autos.describe(include = 'all')

date_crawled         name  seller offer_type       price_$  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year gear_box       power_PS  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  num_photos   postal_code  \
count             299468               371528    371528.0  371528.00000   
unique                 2                  114         NaN           NaN   
top                 nein  2016-04-03 00:00:00         NaN           NaN   
freq              263182                14450         NaN           NaN   
mean                 NaN                  NaN         0.0   50820.66764   
std                  NaN                  NaN         0.0   25799.08247   
min                  NaN                  NaN         0.0    1067.00000   
25%                  NaN                  NaN         0.0   30459.00000   
50%                  NaN                  NaN         0.0   49610.00000   
75%                  NaN                  NaN         0.0   71546.00000   
max                  NaN                  NaN         0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

### Interpreting Data Description

* **count** - shows how many rows have non-missing values.
* **mean** - which is the average.
* **std** - standard deviation, which measures how numerically spread out the values are.
* **min** - smallest value is the min.
* **25%** - quarter way through the list, a number bigger than 25% and smaller than 75% of the values.
* **50th and 75th percentiles** are defined analogously
* **max** - the largest number.

### ⛳Observations 3 - Data Exploration Observations
* **The num_photos column has many 0 as the values in describe table, whether its min or max category. So we need to explore the column further.**
* **Column Seller has one common value for almost all the rows, i.e., 'private'. So we can drop the column.**
* **Similarly Column offer_type has one common value for almost all the rows, i.e., 'Angebot'. So we can drop the column.**
* **Similarly Column abtest can be deleted**

### ⭐Step5 - Dropping unnecessary Columns based on Data Exploration Observations

**✏️Checking num_photos column**

In [9]:
autos["num_photos"].value_counts()

0    371528
Name: num_photos, dtype: int64

**✏️Checking ab_test column**

In [10]:
autos["ab_test"].value_counts()

test       192585
control    178943
Name: ab_test, dtype: int64

### ⛳Observations 4 - Exploring num_photos Observations

All the values are 0, so this column can be deleted from dataset along with previously detected other two.

✏️**Dropping unnecessary Columns**

In [11]:
autos = autos.drop(["num_photos", "seller", "offer_type", "ab_test"], axis=1)

### ⭐Step6 - Detecting Outliers

### Commands used to Analyze the columns and Detecting Outliers

**Outliers** - are unrealistically high or low values that we might want to remove.

We'll use:
* **Series.unique().shape** - to see how many unique values
* **Series.describe()** - to view min/max/median/mean etc
* **Series.value_counts()** - with some variations: eg, chained to .head() if there are lots of values.
* **Series.sort_index()** - with ascending= True or False to view the highest and lowest values with their counts 

✏️**Analyzing kilometer Column**

In [12]:
print("No of unique values in kilometer column:" ,autos["kilometer"].unique().shape)

print("min/max/median/mean values in kilometer column:" ,autos["kilometer"].describe())

print("No of values in kilometer column:" ,autos["kilometer"].value_counts())

print("Highest values with their no of counts in  kilometer column:" ,autos["kilometer"].value_counts().sort_index(ascending = False).head(5))

print("Highest values with their no of counts in kilometer column:" ,autos["kilometer"].value_counts().sort_index(ascending = True).head(5))


No of unique values in kilometer column: (13,)
min/max/median/mean values in kilometer column: count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: kilometer, dtype: float64
No of values in kilometer column: 150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: kilometer, dtype: int64
Highest values with their no of counts in  kilometer column: 150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
Name: kilometer, dtype: int64
Highest values with their no of counts in kilometer column: 5000     7069
10000    1949
20000    5676
30000    6041
40000    6376
Name: kilometer, dtype: int64


### ⛳Observations 5 - Exploring kilometer column Observations
* **There are no outliers, min value is 5000 and max value is 150000, which is practically possible.**
* **Mostly values are rounded off.**
* **Vehicles description uploaded on eBay has good mileage feature.**
* **Overall, we are good to go with the column.**

✏️**Analyzing price_$ Column**

In [13]:
print("No of unique values in price_$ column:" ,autos["price_$"].unique().shape)

print("min/max/median/mean values in price_$ column:" ,autos["price_$"].describe())

print("No of values in price_$ column:" ,autos["price_$"].value_counts())

print("Highest values with their no of counts in price_$ column:" ,autos["price_$"].value_counts().sort_index(ascending=False).head(20))

print("Highest values with their no of counts in price_$ column:",autos["price_$"].value_counts().sort_index(ascending=True).head(20))

No of unique values in price_$ column: (5597,)
min/max/median/mean values in price_$ column: count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price_$, dtype: float64
No of values in price_$ column: 0         10778
500        5670
1500       5394
1000       4649
1200       4594
          ...  
23456         1
171000        1
21830         1
13485         1
8188          1
Name: price_$, Length: 5597, dtype: int64
Highest values with their no of counts in price_$ column: 2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price_$, dtype: int64
Highest values with their

### ⛳Observations 6 - Exploring price_$ column Observations
* **There are 5597 unique values in the price column.**
* **The minimum value for price is 0 that accounts for 10778 times, which is 3% of the cars. Thus we can consider to remove these rows.**
* **The maximum value for price is 2147483647 which is quite high.**

### 🔑Decisions

* **We can remove the rows with zero values, which is only 3% of the cars data.**
* **We can remove rows with very high values such as 2147483647, whose occurence is only one time and a quite big value.**


**We know that eBay is an auction site, where the opening bid is 1 dollar. We can keep the dollar 1 items, but remove anything above $99999999, since it seems that prices increase steadily to that number and then jump up to less realistic numbers.**


### ⭐Step7 - Removing Outliers

In [14]:
#using df[df["col"].between(x,y)] command
autos = autos[autos["price_$"].between(1,99999999)]
autos["price_$"].describe()

count    3.607490e+05
mean     1.185906e+04
std      6.888791e+05
min      1.000000e+00
25%      1.250000e+03
50%      3.000000e+03
75%      7.490000e+03
max      1.000000e+08
Name: price_$, dtype: float64

So we have removed price value 0 and 2147483647 from dataset.

### ⭐Step8 - Exploring the date columns

**There are a number of columns with date information:**

* **date_crawled**
* **registration_month**
* **registration_year**
* **ad_created**
* **last_seen**

**Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values.**

In [15]:
# Exploring each columns
autos[['date_crawled','ad_created','last_seen']][0:4]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17

### ⛳Observations 7 - Date Columns Observations
* **The first 10 characters represent the day (e.g. 2016-03-24).** 
* **To understand the date range, we can extract just the date values, by using Series.value_counts() to generate a distribution, and then sort by the index.**
* **To select the first 10 characters in each column, we can use Series.str[:10]**

✏️**Calculating distribution of the remaining values by using Series.value_counts(normalize=True) command for columns -** 

***'date_crawled','ad_created','last_seen'*** 


In [16]:
# To include missing values in the distribution and to use percentages instead of counts,and  rank by date in ascending order 
(autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index().head(10))


2016-03-05    0.025547
2016-03-06    0.014481
2016-03-07    0.035648
2016-03-08    0.033472
2016-03-09    0.034112
2016-03-10    0.032643
2016-03-11    0.032773
2016-03-12    0.036241
2016-03-13    0.015778
2016-03-14    0.036336
Name: date_crawled, dtype: float64

### ⛳Observations8-  date_crawled column Observations
**Data suggest that site was crawled daily over roughly a one month period in March and April 2016. The distribution of listings crawled on each day is roughly uniform.**

In [17]:
# for ad_created column
(autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index())


2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.039008
2016-04-04    0.037733
2016-04-05    0.011617
2016-04-06    0.003119
2016-04-07    0.001555
Name: ad_created, Length: 114, dtype: float64

In [18]:
autos["ad_created"].str[:10].describe()

count         360749
unique           114
top       2016-04-03
freq           14072
Name: ad_created, dtype: object

### ⛳Observations9 - ad_created column Observations
* **In March month most of the ads were created**

In [19]:
# for last_seen column
(autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index())

2016-03-05    0.001267
2016-03-06    0.004097
2016-03-07    0.005200
2016-03-08    0.007942
2016-03-09    0.009824
2016-03-10    0.011459
2016-03-11    0.012954
2016-03-12    0.023241
2016-03-13    0.008408
2016-03-14    0.012177
2016-03-15    0.016324
2016-03-16    0.016413
2016-03-17    0.028704
2016-03-18    0.006894
2016-03-19    0.016330
2016-03-20    0.019892
2016-03-21    0.020031
2016-03-22    0.020507
2016-03-23    0.018015
2016-03-24    0.019163
2016-03-25    0.018999
2016-03-26    0.015956
2016-03-27    0.016718
2016-03-28    0.022190
2016-03-29    0.023285
2016-03-30    0.023723
2016-03-31    0.024238
2016-04-01    0.023900
2016-04-02    0.024979
2016-04-03    0.025322
2016-04-04    0.025527
2016-04-05    0.126966
2016-04-06    0.218936
2016-04-07    0.130418
Name: last_seen, dtype: float64

In [20]:
autos["last_seen"].str[:10].describe()

count         360749
unique            34
top       2016-04-06
freq           78981
Name: last_seen, dtype: object

### ⛳Observations10 -  last_seen column Observations

* Last seen column suggests the last day when listing was viewed which means most propably the day when listing of car was removed or in other words, its the day when the car is sold.
* April month has high percentage in last seen columns which suggest may be sale was done or crawling period got over.
* For now, we are good to go with the data.

✏️**Exploring *registration_year* column**

In [21]:
# for registration_year column
autos["registration_year"].describe()

count    360749.000000
mean       2004.450546
std          82.104648
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

### ⛳Observations11 -  registration_year column Observations
**registration_year column contains some odd values:**

* **The minimum value is 1000, before cars were invented**
* **The maximum value is 9999, many years into the future**


* **Because a car can't be first registered after the listing was seen, any vehicle with a registration year above 2016 is definitely inaccurate.**
* **Determining the earliest valid year is more difficult. Realistically, it could be somewhere in the first few decades of the 1900s.**


✏️**Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.**

In [22]:
# percentage of our data with invalid values 
(autos["registration_year"].between(1900,2016)).sum() / autos.shape[0]

0.9612362057829682

In [23]:
# counting the number of listings with cars that fall outside the 1900 - 2016 interval
autos["registration_year"].value_counts(normalize=True)

2000    0.064125
1999    0.061078
2005    0.060233
2006    0.055465
2001    0.054742
          ...   
3200    0.000003
7800    0.000003
1400    0.000003
1911    0.000003
4100    0.000003
Name: registration_year, Length: 145, dtype: float64

✏️**Removing Outlier rows**

In [24]:
#using df[df["col"].between(x,y)] command
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].describe()

count    346765.000000
mean       2002.894326
std           7.251214
min        1910.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        2016.000000
Name: registration_year, dtype: float64

Thus data is between 1900 - 2016 years of registration.

✏️**Exploring *registration_month* column**

In [25]:
# for registration_month column
autos["registration_month"].describe()

count    346765.000000
mean          5.835485
std           3.665904
min           0.000000
25%           3.000000
50%           6.000000
75%           9.000000
max          12.000000
Name: registration_month, dtype: float64

In [26]:
autos["registration_month"].value_counts()

3     34478
6     31418
0     29734
4     29305
5     29030
7     27529
10    26032
11    24282
12    24003
9     23925
1     23272
8     22493
2     21264
Name: registration_month, dtype: int64

### ⛳Observations11 -  registration_month column Observations

* Around 8 % of car data has 0 value as month of registration. This might happen because of not recording month of registration in several entries.
* In March and June, more registrations were made.
* The distribution of registrations in other months are roughly uniform.
* Either we can remove the data with moth value as 0 which contributes to 8% or can go with it due to its percentage.

### ⭐Step9 -Exploring Price by Brand Column

In [27]:
autos["brand"].value_counts(normalize=True)

volkswagen        0.211697
bmw               0.109870
opel              0.106386
mercedes_benz     0.096832
audi              0.089539
ford              0.068911
renault           0.047502
peugeot           0.030144
fiat              0.025686
seat              0.018655
skoda             0.015682
mazda             0.015382
smart             0.014327
citroen           0.013949
nissan            0.013594
toyota            0.012928
hyundai           0.009969
sonstige_autos    0.009557
mini              0.009381
volvo             0.009145
mitsubishi        0.008233
honda             0.007530
kia               0.006912
suzuki            0.006365
alfa_romeo        0.006310
porsche           0.006261
chevrolet         0.005024
chrysler          0.003861
dacia             0.002494
jeep              0.002195
land_rover        0.002166
daihatsu          0.002160
subaru            0.002117
jaguar            0.001736
saab              0.001465
daewoo            0.001456
trabant           0.001410
l

In [28]:
autos["brand"].describe()

count         346765
unique            40
top       volkswagen
freq           73409
Name: brand, dtype: object

### ⛳Observations12 -  brand column Observations
Volkswagen is by far the most popular brand.

In [29]:
# Checking top popular brands
brand_counts = autos["brand"].value_counts(normalize=True)
popular_brands = brand_counts[brand_counts > .05].index
print(popular_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


**Top Popular Brands are - 'volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'**
**Let's check these brands by their prices using aggregation technique.**

Method - 


- Identify the unique values we want to aggregate by
- Create an empty dictionary to store our aggregate data
- Loop over the unique values, and for each:
    - Subset the dataframe by the unique values
    - Calculate the mean of whichever column we're interested in
    - Assign the val/mean to the dict as k/v.


In [30]:
brand_mean_prices = {}

for brand in popular_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_price = brand_only["price_$"].mean()
    brand_mean_prices[brand] = int(mean_price)

brand_mean_prices

{'volkswagen': 14083,
 'bmw': 15201,
 'opel': 3363,
 'mercedes_benz': 17959,
 'audi': 16651,
 'ford': 8945}

Using pandas series and dataframe constructor Methods

In [31]:
bmp_series = pd.Series(brand_mean_prices)
print(bmp_series)

volkswagen       14083
bmw              15201
opel              3363
mercedes_benz    17959
audi             16651
ford              8945
dtype: int64


In [32]:
df = pd.DataFrame(bmp_series, columns=['mean_price'])
df

mean_price
volkswagen          14083
bmw                 15201
opel                 3363
mercedes_benz       17959
audi                16651
ford                 8945

### ⛳Observations13 -  brand by price_$ Observations

> Table depicting Brand and Price Association

| Brand         | Price                         |
| -----         | -----------                   |
| Audi          | more expensive                |
| BMW           | more expensive                |                
| Mercedes Benz | more expensive                |
| Volkswagen    | Optimum / Popular Choice      |
| Ford          | less expensive                |
| Opel          | less expensive                |


### ⭐Step10 -Exploring Mileage by mean_price

In [34]:
brand_mean_mileage = {}

for brand in popular_brands:
    brand_only = autos[autos["brand"] == brand]
    mean_mileage = brand_only["kilometer"].mean()
    brand_mean_mileage[brand] = int(mean_mileage)

mean_mileage = pd.Series(brand_mean_mileage).sort_values(ascending=False)
mean_prices = pd.Series(brand_mean_prices).sort_values(ascending=False)


In [35]:
brand_mileage = pd.DataFrame(mean_mileage,columns=['mean_mileage'])
brand_mileage

mean_mileage
bmw                  132790
mercedes_benz        130568
audi                 129436
opel                 128721
volkswagen           128385
ford                 123662

In [36]:
brand_mileage["mean_price"] = mean_prices
brand_mileage

mean_mileage  mean_price
bmw                  132790       15201
mercedes_benz        130568       17959
audi                 129436       16651
opel                 128721        3363
volkswagen           128385       14083
ford                 123662        8945

### ⛳Observations13 -  Exploring Mileage by mean_price Observations
**Mileage doesn't change much with the Price. However there is a slight trend of higher mileage for higher price and vice versa.**


### ⭐Step10 - Data cleaning next steps
**Identify categorical data that uses german words, translate them and map the values to their english counterparts**

In [37]:
# to print all column labels
autos.columns

Index(['date_crawled', 'name', 'price_$', 'vehicle_type', 'registration_year',
       'gear_box', 'power_PS', 'model', 'kilometer', 'registration_month',
       'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 'postal_code',
       'last_seen'],
      dtype='object')

In [38]:
# Finding German words in vehicle_type column
autos["vehicle_type"].unique()

array([nan, 'coupe', 'suv', 'kleinwagen', 'limousine', 'cabrio', 'bus',
       'kombi', 'andere'], dtype=object)

In [39]:
# Translating German Values to English by Mapping Technique
vehicle_type = autos["vehicle_type"]
map_vehicle = {
    "limousine":"limousine",
    "kleinwagen":"small car",
    "kombi":"combi",
    "bus": "bus",
    "cabrio":"convertible",
    "coupe":"coupe",
    "suv":"suv",
    "andere":"other"
}

autos["vehicle_type"] = pd.Series(vehicle_type).map(map_vehicle)

In [40]:
autos["gear_box"].unique()

array(['manuell', 'automatik', nan], dtype=object)

In [41]:
# Translating German Values to English by Mapping Technique
gear_box = autos["gear_box"]
map_gear_box = {
    "manuell":"manual",
    "automatik":"automatic",
    }

autos["gear_box"] = pd.Series(gear_box).map(map_gear_box)

In [42]:
# for fuel_type
autos["fuel_type"].unique()

array(['benzin', 'diesel', nan, 'lpg', 'andere', 'hybrid', 'cng',
       'elektro'], dtype=object)

In [43]:
# Translating German Values to English by Mapping Technique
fuel_type = autos["fuel_type"]
map_fuel_type = {
    "benzin" :"petrol",
    "diesel":"diesel",
    "lpg" :"lpg",
    "cpg":"cpg",
    "hybrid" :"hybrid",
    "andere":"other",
    "elektro":"electric",
}

autos["fuel_type"] = pd.Series(fuel_type).map(map_fuel_type)

In [44]:
# for unrepaired_damage
autos["unrepaired_damage"].unique()

array([nan, 'ja', 'nein'], dtype=object)

In [45]:
# Translating German Values to English by Mapping Technique
damage_type = autos["unrepaired_damage"]
map_damage_type = {
    "nein" :"no",
    "ja":"yes",
}

autos["unrepaired_damage"] = pd.Series(damage_type).map(map_damage_type)

### ⛳Observations14 -
**Rest of the columns dont haave German Words, either they are numeric or name of model or Brand**

### ⭐Step11 - Data cleaning next steps
**Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321**

In [46]:
#to convert in numeric type
autos["date_crawled"] = autos["date_crawled"].str[:10].str.replace("-","").astype(int)
print(autos["date_crawled"].head())

0    20160324
1    20160324
2    20160314
3    20160317
4    20160331
Name: date_crawled, dtype: int32


In [47]:
#to convert in numeric type
autos["last_seen"] = autos["last_seen"].str[:10].str.replace("-","").astype(int)
print(autos["last_seen"].head())

0    20160407
1    20160407
2    20160405
3    20160317
4    20160406
Name: last_seen, dtype: int32


### ⛳Observations14 -
Now all columns of dates are uniform numeric data

### ⭐Step12 - Data cleaning next steps
**See if there are particular keywords in the name column that you can extract as new columns**

In [48]:
autos["name"].unique()

array(['Golf_3_1.6', 'A5_Sportback_2.7_Tdi',
       'Jeep_Grand_Cherokee_"Overland"', ...,
       'Smart_smart_leistungssteigerung_100ps', 'VW_Golf_Kombi_1_9l_TDI',
       'BMW_M135i_vollausgestattet_NP_52.720____Euro'], dtype=object)

In [49]:
autos["name"].value_counts()

BMW_318i                                                          627
Ford_Fiesta                                                       620
Volkswagen_Golf_1.4                                               603
Opel_Corsa                                                        599
BMW_316i                                                          523
                                                                 ... 
Fahrbereiter_renault_clio_zum_ausschlachten                         1
Opel_Adam_1.2_Open_Air                                              1
Mercedes_Benz_M_Klasse_ML_280_CDI                                   1
VW_T4_allstar_2_4_l__AAB_Bus_Bastlerfahrzeug                        1
CL_500____der_Bolide_aus_der_L.A._Schmiede_"_Einer_von_35_Stck      1
Name: name, Length: 214096, dtype: int64

In [50]:
s = pd.Series(autos["name"])


In [51]:
s.str.rsplit(pat = "_", n=2)

0                                           [Golf, 3, 1.6]
1                                 [A5_Sportback, 2.7, Tdi]
2                       [Jeep_Grand, Cherokee, "Overland"]
3                                   [GOLF_4_1_4, , 3TÜRER]
4                       [Skoda_Fabia_1.4_TDI, PD, Classic]
                                ...                       
371523                      [Suche_t4___vito_ab, 6, sitze]
371524           [Smart_smart, leistungssteigerung, 100ps]
371525              [Volkswagen_Multivan_T4_TDI, 7DC, UY2]
371526                          [VW_Golf_Kombi_1, 9l, TDI]
371527    [BMW_M135i_vollausgestattet_NP_52.720__, , Euro]
Name: name, Length: 346765, dtype: object

**Thus, name column can be split into Model name and Engine or other Specification columns using extract and split commands**

### ⭐Step13 - Data Analysis next steps
**Find the most common brand/model combinations**

In [52]:
# grouping the dataset by model and brand
brand_mod_group = autos.groupby(["brand","model"])
# find the count and sort descending
brand_model = brand_mod_group["date_crawled"].count().sort_values(ascending=False)
print(brand_model[:10])

brand          model   
volkswagen     golf        27554
bmw            3er         19472
volkswagen     polo        12057
opel           corsa       11598
               astra       10045
audi           a4           9802
volkswagen     passat       9724
mercedes_benz  c_klasse     8444
bmw            5er          8208
mercedes_benz  e_klasse     7261
Name: date_crawled, dtype: int64


### ⛳Observations15 -  

> Table depicting common brand/model combinations

| Brand         | Model                         |
| -----         | -----------                   |
| volkswagen    | golf, polo , passat           |
| BMW           | 3er                           |                
| Mercedes Benz | c_klasse, e_klasse            |
| Opel          | corsa,  astra,  a4            |

### ⭐Step14 - Data Analysis next steps
**Split the kilometer into groups, and use aggregation to see if average prices follows any patterns based on the milage.**

In [53]:
autos["kilometer"] = pd.qcut(autos['kilometer'], q=10,duplicates ="drop")

odometri = autos["kilometer"].unique()
print(odometri)
avg_price_for_groups_km ={}

for od in odometri:
    sel_rows = autos[autos["kilometer"] == od]
    mean_km_price_for_groups = sel_rows["price_$"].mean()
    odo = int(mean_km_price_for_groups)
    avg_price_for_groups_km[od] = odo
    

odo_series = pd.Series(avg_price_for_groups_km)

df3 =pd.DataFrame(odo_series,columns=["mean_price"])
print(df3)

[(125000.0, 150000.0], (90000.0, 125000.0], (60000.0, 90000.0], (4999.999, 60000.0]]
Categories (4, interval[float64]): [(4999.999, 60000.0] < (60000.0, 90000.0] < (90000.0, 125000.0] < (125000.0, 150000.0]]
                      mean_price
(125000.0, 150000.0]        9644
(90000.0, 125000.0]         7992
(60000.0, 90000.0]         10095
(4999.999, 60000.0]        29199


### ⛳Observations16 -  
The kilometer column has been split in 4 groups: 5.000-60.000 , 60.000-90.000, 90.000-125.000, 125.000-150.000

Higher the price, Higher the mileage and vice versa

### ⭐Step15 - Data Analysis next steps
**How much cheaper are cars with damage than their non-damaged counterparts?**

In [55]:
autos["unrepaired_damage"].value_counts()

no     252827
yes     33196
Name: unrepaired_damage, dtype: int64

In [56]:
repaired_unrepaired_price = {}

damage = autos["unrepaired_damage"].unique()
print(damage)

for d in damage:
    selected_rows = autos[autos["unrepaired_damage"] == d]
    mean = selected_rows["price_$"].mean()
    repaired_unrepaired_price[d] = float(mean)

print(repaired_unrepaired_price)

[nan 'yes' 'no']
{nan: nan, 'yes': 6685.262230389203, 'no': 9335.89641929066}


### ⛳Observations17 -
**Cars with unrepaired damage costs less, while cars with repairable damage cost less**

# 🔦CONCLUSION
**We have done cleaning in dataset and performed analysis by comparing features with each other. We conclude tat Volkswagen is more popular car, times of sales spike using date_crawled, last_seen columns, etc. We also conclude relation between price, model and mileage.We can build a Machine Learning Model also in this project to predict price by using ML Models, Feature Engineering, etc**